homeassistant/components/harmony/subscriber.py

```diff
@@ -0,0 +1,98 @@
+"""Mixin class for handling harmony callback subscriptions."""
+
+import asyncio
+import logging
+from typing import Any, Callable, NamedTuple, Optional
+
+from homeassistant.core import callback
+
+_LOGGER = logging.getLogger(__name__)
+
+NoParamCallback = Optional[Callable[[object], Any]]
+ActivityCallback = Optional[Callable[[object, tuple], Any]]
+
+
+class HarmonyCallback(NamedTuple):
+    """Callback type for Harmony Hub notifications."""
+
+    connected: NoParamCallback
+    disconnected: NoParamCallback
+    config_updated: NoParamCallback
+    activity_starting: ActivityCallback
+    activity_started: ActivityCallback
+
+
+class HarmonySubscriberMixin:
+    """Base implementation for a subscriber."""
+
+    def __init__(self):
+        """Initialize an subscriber."""
+        super().__init__()
+        self._subscriptions = []
+
+    @callback
+    def async_subscribe(self, update_callback: HarmonyCallback) -> Callable:
+        """Add a callback subscriber."""
+        self._subscriptions.append(update_callback)
```

Comment:
Wrap these in `HassJob` and then call `hass.async_run_job` on the jobs in the array.

Reply:
I rejiggered this a little bit. The callbacks within this HarmonyCallback structure are optional, so not necessarily all of them will be present. Instead of doing logic twice to check for their existence (once while wrapping, and again when calling later on) or keeping several sets of callbacks I used `async_run_job` directly. I don't think the lack of memoization will really affect performance that much.

Thank you for suggesting this - this path is much more convenient compared to manually handling async vs standard function calls.